PHPUnit 11

greg0ire opened the pull request:
This upgrades to PHPUnit 11 _and_ addresses all newly emitted deprecations.

greg0ire committed:
Migrate away from getMockForAbstractClass()

It has been deprecated.

greg0ire committed:
Migrate to willReturn()

self::returnValue() and self::onConsecutiveCalls() has been deprecated.

greg0ire committed:
Migrate away from assertStringNotMatchesFormat()

It has been deprecated.

greg0ire committed:
Migrate away from annotations in tests

greg0ire committed:
Use PHPUnit 11 when possible

greg0ire commented:
:exploding_head: I'm reproducing the failure on the main branch

xabbuh commented on tests/Tests/ORM/Functional/PersistentCollectionTest.php:
```diff
@@ -31,6 +31,15 @@ protected function setUp(): void
 
+    protected function tearDown(): void
+    {
+        if (! class_exists(PersistentObject::class)) {
```
Is this really required? https://github.com/doctrine/orm/blob/31e6a1da87fc76dbc7273b96f9a8df6845bd7c95/tests/Tests/OrmFunctionalTestCase.php#L542-L545 does already handle this case gracefully, doesn't it?

greg0ire replied:
Ah I didn't see this. But no, it no longer handles this gracefully when running the whole test suite, so maybe sharedConn now persists between tests.

greg0ire replied:
@xabbuh I debugged it, and `sharedConn` is always set. 

Moreover, using this code:

```
    protected function tearDown(): void
    {
        try {
            parent::tearDown();
        } catch (\Throwable $e) {
            var_dump($e->getMessage());
        }
    }
```

With PHPUnit _10_
prints this:

```
string(41) "Call to a member function clear() on null"
```

So the same issue exists with both versions of PHPUnit, it's just that for some reason, PHPUnit 10 swallows the exception.

greg0ire replied:
I replaced the commit that introduce this change with another that acts at the `OrmFunctionalTestCase` level, by also checking `$this->_em`.

xabbuh replied:
Ah that explains it probably: sebastianbergmann/phpunit#6281

greg0ire replied:
Yes, nice find!

greg0ire committed:
Check extra condition to decide if a test was skipped

It seems that this could happen with PHPUnit 10, then tearDown() would
crash when calling `clear()` on null, but then PHPUnit 10 did not show
that exception.
